### PR TITLE
Music Plugins

### PR DESCRIPTION
**Pull request by tanndlin**
Only Moderators are allowed to use the plugins

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,16 +24,21 @@
...
+    "ffmpeg-static": "^4.3.0",
     "moment": "^2.29.1",
     "mongodb": "^3.5.9",
+    "node-opus": "^0.3.3",
     "typescript": "^4.0.5",
     "winston": "^3.3.3",
-    "winston-papertrail": "^1.0.5"
+    "winston-papertrail": "^1.0.5",
+    "yt-search": "^2.8.0",
+    "ytdl-core": "^4.5.0"
```
Comment (c650) [inline package.json:35]: is this stuff against yt terms of service?

File: src/app/jobs/music.job.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { IContainer } from '../../common/types';
+import { Job } from '../../common/job';
+
+export class MusicJob extends Job {
+  public interval: number = 1000 * 60 * 5; //Every 5 minute
```
Comment (c650) [inline src/app/jobs/music.job.ts:5]: pls use the duration thingy (`moment`)

File: src/app/jobs/music.job.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  constructor() {
+    super();
+  }
+
+  public async execute(container: IContainer) {
+    if (!container.musicService.isInVC()) {
+      return;
+    }
+
+    const isSilent = container.musicService.isStreaming();
```
Comment (c650) [inline src/app/jobs/music.job.ts:19]: how are we silent if the service is streaming?

File: src/app/plugins/queue.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { Plugin } from '../../common/plugin';
+import { IContainer, IMessage, ChannelType } from '../../common/types';
+import { MemberUtils } from '../util/member.util';
+
+export class QueuePlugin extends Plugin {
+  public name: string = 'Play Plugin';
+  public description: string = 'Plays music in a voicechannel';
+  public usage: string = 'play <song name>';
```
Comment (c650) [inline src/app/plugins/queue.plugin.ts:8]: copypasta error?

File: src/app/plugins/skip.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { Plugin } from '../../common/plugin';
+import { IContainer, IMessage, ChannelType } from '../../common/types';
+import { MemberUtils } from '../util/member.util';
+
+export class SkipPlugin extends Plugin {
+  public name: string = 'Play Plugin';
+  public description: string = 'Plays music in a voicechannel';
+  public usage: string = 'play <song name>';
+  public pluginAlias = [];
```
Comment (c650) [inline src/app/plugins/skip.plugin.ts:9]: saa

File: src/app/plugins/skip.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  public async execute(message: IMessage, args: string[]) {
+    if (!message.member) {
+      return;
+    }
+
+    const isModerator = MemberUtils.hasRole(message.member, 'moderator');
+    if (!isModerator) {
+      await message.reply('You must be a moderator to use this plugin.');
+      return;
+    }
```
Comment (c650) [inline src/app/plugins/skip.plugin.ts:25]: moar copypasta

File: src/app/plugins/skip.plugin.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    const ret = await this.container.musicService.skip();
```
Comment (c650) [inline src/app/plugins/skip.plugin.ts:27]: `ret` not descriptive

File: src/services/music.service.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+  public async queue(message: IMessage, songName: string): Promise<string> {
+    if (!message.member) {
+      return 'Could not resolve to user';
+    }
+
+    const isModerator = MemberUtils.hasRole(message.member, 'moderator');
+    if (!isModerator) {
+      return 'You must be a moderator to use this plugin';
+    }
```
Comment (c650) [inline src/services/music.service.ts:24]: if we do this here, why also do in plugin?

File: src/services/music.service.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+    //Should make sure the bot isnt occupied
```
Comment (c650) [inline src/services/music.service.ts:26]: nit: `//<space>`

also, wym bot not occupied here?

File: src/services/music.service.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    const targetVC = message.member.voice.channel;
+    if (!targetVC) {
+      return 'You must be in a voice channel to use this plugin';
+    }
+
+    if (this._currentVoiceChannel && targetVC.name !== this._currentVoiceChannel.name) {
+      return 'I can only be used in one voice channel at a time';
+    }
```
Comment (c650) [inline src/services/music.service.ts:34]: the earlier commend makes more sense here..

File: src/services/music.service.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    //Send message in bot_commands what song is starting
+    const botChan = this._guildService.getChannel('bot_commands') as TextChannel;
+    await botChan.send(this._videoToEmbed(video));
+
+    this._connection = await vc.join();
+    this._currentVoiceChannel = vc;
+
+    //Get Audio Stream
+    const stream = ytdl.default(video.url, { filter: 'audioonly', highWaterMark: 1 << 25 });
```
Comment (c650) [inline src/services/music.service.ts:88]: what does `highWaterMark` do

File: src/services/music.service.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      //Otherwise, play next song
+      this._play(vc);
+    });
+  }
+
+  private _getDurationOfQueue(): string {
+    const seconds = this._queue.reduce((acc, cur) => acc + cur.seconds, 0);
+    const duration = moment.duration(seconds, 'seconds');
+    return `${duration.minutes()} minutes ${duration.seconds()} seconds`;
+  }
```
Comment (c650) [inline src/services/music.service.ts:109]: lit <3

File: src/services/music.service.ts [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+      .map((song, i) => `**${i + 1}** - [${song.title}](${song.url})`)
+      .join('\n');
+
+    const embed = new MessageEmbed();
+    embed.setTitle(`${this._queue.length} Songs in Queue`);
+    embed.addField('Tracks', trackList || 'No Songs Currently in Queue');
+
+    embed.setFooter(`Queue Duration: ${this._getDurationOfQueue()}`);
+    return embed;
+  }
```
Comment (c650) [inline src/services/music.service.ts:144]: sooo clean i love it

File: src/services/music.service.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+  public isInVC(): boolean {
+    return Boolean(this._currentVoiceChannel);
+  }
+
+  public leaveVC(): void {
+    if (!this._currentVoiceChannel) {
+      return;
+    }
+
+    this._currentVoiceChannel.leave();
```
Comment (c650) [inline src/services/music.service.ts:159]: `this._currentVoiceChannel?.leave`

the ? is the "try operator"

**Review by c650 — COMMENTED** [review]
i never particularly wanted to have Lion do music stuff because: 1. just get spotify; 2. kind of annoying in voice (if its on when im in voice i usually just server mute the bot); 3. youtube tos

File: package.json (hunk 1 above)
Comment (tanndlin) [reply package.json:35]: didnt think about that, after reading about it, it seems very questionable Sadge
never thought about that

**Comment by tanndlin** [comment]
welp, even if we close, this. I did learn that bots can hyperlink. pog

**Comment by c650** [comment]
> welp, even if we close, this. I did learn that bots can hyperlink. pog

we can sleep on it for a bit. Also if you check out how other bots handle the legal stuff, it would make me more comfortable merging it

**Comment by tanndlin** [comment]
Closing but keeping my branch

**Pull request closed without merging**